xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -386,4 +388,60 @@ void reportClientStats(String clusterName, String serverUri, LoadStatsStore load
    */
   void cancelClientStatsReport(String clusterName) {
   }
+
+  /**
+   * An {@link ObjectPool} holding reference and ref-count of an {@link XdsClient} instance.
+   * Once the ref-count gets back to zero, the xds client will be shutdown and later calls of {@lint
+   * #getObject()} will return null.
+   */
+  static final class RefCountedXdsClientObjectPool implements ObjectPool<XdsClient> {
+    private final Object lock = new Object();
+
+    @Nullable
+    private XdsClient xdsClient;
+    private int refCount;
+
+    RefCountedXdsClientObjectPool(XdsClient xdsClient) {
+      this.xdsClient = Preconditions.checkNotNull(xdsClient, "xdsClient");
+    }
+
+    @CheckReturnValue
+    @Nullable
+    @Override
+    public XdsClient getObject() {
+      synchronized (lock) {
+        if (xdsClient != null) {
```

Comment:
If xdsClient is null, throwing seems like a good idea. Or are you expecting that the caller would see the `null` and then create their own xds client?

Reply:
I'm worried in the case resolver shutdown too early before a child balancer `handleResolvedAddresses()`, and when `handleResolvedAddresses()` it should get null instead of throw, and then the child balancer can shutdown gracefully.

The caller seeing `getObject()` returning `null` will not create their own because the Attribute is not `null`.

Reply:
Hmm... Both options seem _slightly_ off. I'm thinking now it may be better to have the RefCountedXdsClientObjectPool create the XdsClient directly, instead of having it passed in. XdsClient would then be passed a factory or the arguments necessary to create the XdsClient. So this would mean any time the reference count is 0, `xdsClient == null`. And it would mean it could handle the reference count transitioning from 0→1 multiple times (by creating multiple XdsClients over time).

Reply:
Done with the "have the RefCountedXdsClientObjectPool create the XdsClient directly" approach.